File: gfg/2024_6_june/c++/09_convert_array_into_zig_zag_fashion.cpp
// NOTE
// We don't need to sort the array, this problem can be solved in O(n).
// We just need to make sure that at an even position value is smaller than its 
// right and at an odd position value is greater than its right.

// TIME COMPLEXITY
// O(logn)

// SPACE COMPLEXITY
// O(1)

// CODE

class Solution {
    
  public:
    void zigZag(int n, vector<int> &arr) {
        for(int i = 0; i < arr.size() - 1; i++){
            // even and ith > (i +1)th  
            if(i % 2 == 0 && arr[i] > arr[i+1]){
                swap(arr[i],arr[i+1]);
            }
            // odd and ith < (i +1)th  
            if(i % 2 != 0 && arr[i] < arr[i+1]){
                swap(arr[i],arr[i+1]);
            }
        }
    }
};

File: gfg/2024_6_june/python/09_convert_array_into_zig_zag_fashion.py
# NOTE
# We don't need to sort the array, this problem can be solved in O(n).
# We just need to make sure that at an even position value is smaller than its 
# right and at an odd position value is greater than its right.

# CODE

from typing import List
class Solution:
    def zigZag(self, n : int, arr : List[int]) -> None:
        for i in range(n - 1):
            if i % 2 == 0:
                if arr[i] > arr[i + 1]:
                    arr[i], arr[i + 1] = arr[i + 1], arr[i]
            else:
                if arr[i] < arr[i + 1]:
                    arr[i], arr[i + 1] = arr[i + 1], arr[i]